### PR TITLE
fix: avoid modules_to_save key collisions with adapter prefixes

### LINKED ISSUE #3433: modules_to_save module whose name contains the LoRA prefix ("lora_") is silently not restored on load
### System Info

peft `main` @ `cea8213` (0.19.2.dev0), torch 2.13, transformers 5.13, CPU-only.

### Who can help?

This is a state-dict save/load correctness issue in `modules_to_save` handling.

### Reproduction

A `modules_to_save` module whose name contains the adapter's parameter prefix (for LoRA, `lora_`) is silently not restored on load. The clearest example is a module literally named `lora_head`:

```python
import copy, torch
from torch import nn
from peft import LoraConfig, get_peft_model, get_peft_model_state_dict, set_peft_model_state_dict

class ModelWithLoraNamedHead(nn.Module):
    def __init__(self):
        super().__init__()
        self.body = nn.Linear(2, 2, bias=False)
        self.lora_head = nn.Linear(2, 2, bias=False)   # ordinary module, name contains "lora_"
    def forward(self, x):
        return self.lora_head(self.body(x))

config = LoraConfig(target_modules=["body"], modules_to_save=["lora_head"])
source = get_peft_model(ModelWithLoraNamedHead(), config)
with torch.no_grad():
    source.base_model.model.lora_head.modules_to_save["default"].weight.fill_(7.0)

state_dict = get_peft_model_state_dict(source)
target = get_peft_model(ModelWithLoraNamedHead(), copy.deepcopy(config))
result = set_peft_model_state_dict(target, state_dict)

print("unexpected keys:", [k for k in result.unexpected_keys if "lora_head" in k])
loaded = target.base_model.model.lora_head.modules_to_save["default"].weight
print("weight restored:", bool(torch.allclose(loaded, torch.full_like(loaded, 7.0))))
```

Output:

```
unexpected keys: [
  'base_model.model.lora_head.default.original_module.weight',
  'base_model.model.lora_head.default.modules_to_save.weight',
  'base_model.model.lora_head.default.modules_to_save.default.weight',
]
weight restored: False
```

The trained `lora_head` weight stays at its initialization, and because loading is non-strict the mismatch is silent — a caller that doesn't inspect `result.unexpected_keys` keeps running with the wrong head weights.

### Expected behavior

The `modules_to_save` weight should round-trip regardless of the module's name, as it does for any name that doesn't contain `lora_`.

### Root cause

In `src/peft/utils/save_and_load.py`, `_insert_adapter_name_into_state_dict` decides whether a key belongs to the adapter with a substring test:

```python
if parameter_prefix in key:   # parameter_prefix == "lora_"
```

For a `modules_to_save` module named `lora_head`, `"lora_" in "...lora_head..."` is `True`, so the key is treated as a LoRA parameter and the adapter name is spliced in at the wrong position. A prefix/boundary-aware check (rather than an unanchored `in`) would classify these keys correctly.

### Notes

I came across this while auditing `save_and_load.py` for edge cases; it needs a module name that contains the `lora_` prefix, so it's narrow in practice, but the result is a silent wrong-weight load. I have a small fix along the lines above plus a regression test ready. Per `.ai/AGENTS.md`, I wanted to check in first rather than open a PR directly — would you welcome one? (Disclosure: the audit and this write-up were done with AI assistance; the reproduction above was run against the real editable checkout.)

### PR DESCRIPTION
**Pull request by Solaris-star**
## Summary

Fixes #3433.

A `modules_to_save` module whose name contains a tuner prefix (e.g. `lora_head` under LoRA) is silently not restored on load. Root cause: unanchored substring matching (`parameter_prefix in key`) in:

1. **`_insert_adapter_name_into_state_dict` (load)** — treats `lora_head.weight` as if it were a LoRA leaf (`lora_A.weight`) and rewrites it incorrectly
2. **`get_peft_model_state_dict` (save)** — prefix match over-selects the module, then the wrapper path must re-insert the correct keys; also embedding saves used the same unanchored match against siblings like `embed_tokens_extra`

## Fix

- On load: if the remainder after the prefix still contains `.`, leave the key untouched (adapter params are leaves)
- On save: drop stale prefix-matched keys for `AuxiliaryTrainingWrapper` modules before re-inserting the correct `modules_to_save` keys
- On embedding save: anchor with `k.startswith(f"{embedding_module_name}.")`

## Tests

```text
python -m pytest tests/test_other.py -k ModulesToSaveNameContainsPrefixBug -q
# 2 passed
```

Covers the issue repro (`lora_head` round-trip) and the embedding sibling boundary case.

## Note

A previous community PR (#3451) explored the same fix but was closed for process/coordination reasons while the issue remains open. Happy to adjust if maintainers prefer a different approach.

### CODE REVIEW

File: tests/test_other.py [hunk 1]
```diff
@@ -427,6 +427,62 @@ def test_load_merged_and_unmerged_same_weights(self, path_merged_and_unmerged):
...
+    def test_save_load_round_trip_restores_weight(self):
+        config = LoraConfig(target_modules=["body"], modules_to_save=["lora_head"])
+        source = get_peft_model(self.get_model(), config)
+        with torch.no_grad():
+            source.base_model.model.lora_head.modules_to_save["default"].weight.fill_(7.0)
+
+        state_dict = get_peft_model_state_dict(source)
+        assert {k for k in state_dict if "lora_head" in k} == {"base_model.model.lora_head.weight"}
+
+        target = get_peft_model(self.get_model(), copy.deepcopy(config))
```
Comment (BenjaminBossan) [inline tests/test_other.py:458]: What is the point of deepcopying the config here? PEFT can mutate the config, yes, but if that's the reason, it's already too late here, you'd have to deepcopy the config on the first `get_peft_model` call.

File: tests/test_other.py [hunk 2]
```diff
@@ -427,6 +427,62 @@ def test_load_merged_and_unmerged_same_weights(self, path_merged_and_unmerged):
...
+        return ModelWithLoraNamedModule()
+
+    def test_save_load_round_trip_restores_weight(self):
+        config = LoraConfig(target_modules=["body"], modules_to_save=["lora_head"])
+        source = get_peft_model(self.get_model(), config)
+        with torch.no_grad():
+            source.base_model.model.lora_head.modules_to_save["default"].weight.fill_(7.0)
+
+        state_dict = get_peft_model_state_dict(source)
+        assert {k for k in state_dict if "lora_head" in k} == {"base_model.model.lora_head.weight"}
```
Comment (BenjaminBossan) [inline tests/test_other.py:456]: for clarification, let's add a comment here of how this failed before the fix

File: src/peft/utils/save_and_load.py [hunk 3]
```diff
@@ -384,6 +384,11 @@ def renamed_dora_weights(k):
             module_state_dict = {
                 k.removeprefix(f"{name}."): v for k, v in state_dict.items() if k.startswith(f"{name}.")
             }
+            # Drop keys that were selected only because the module name contains a
+            # tuner prefix (e.g. "lora_head.weight" matched by prefix "lora_"). The
+            # wrapper then re-inserts the correct modules_to_save keys (#3433).
+            for stale_key in [k for k in to_return if k.startswith(f"{name}.")]:
```
Comment (BenjaminBossan) [inline src/peft/utils/save_and_load.py:390]: Honestly, I'm not stoked that we first add the wrong key and then remove it later. And even then, only if the layer happens to be an auxiliary module. Let's take this test:

```python
    def test_no_stray_module_included(self):
        # the base model happens to contain a module whose name starts with "lora_", it should not be part of the state dict
        model = self.get_model()
        config = LoraConfig(target_modules=["body"], modules_to_save=["lora_head"])
        model = get_peft_model(model, config)
        state_dict = get_peft_model_state_dict(model)
        assert not any("lora_head" in k for k in state_dict)
```

It fails even with your fix. I would much rather see a principled fix that prevents wrong modules from being added in the first place. This would probably require changes in many places, which wouldn't be good. So instead, let's do a separate pass through the model and remove those stray keys, independent of auxiliary modules.

File: src/peft/utils/save_and_load.py [hunk 4]
```diff
@@ -543,6 +553,13 @@ def _insert_adapter_name_into_state_dict(
             _, _, suffix = key.rpartition(parameter_prefix)
             if "." in suffix:
                 suffix_to_replace = ".".join(suffix.split(".")[1:])
+                # Adapter parameters are leaf attributes (e.g. "lora_A.weight"). A longer
+                # suffix means the prefix matched a module name that happens to contain
+                # the adapter prefix (e.g. "lora_head.weight" with prefix "lora_"), not
+                # a real adapter parameter — leave those keys untouched (#3433).
+                if "." in suffix_to_replace:
+                    peft_model_state_dict[key] = val
```
Comment (BenjaminBossan) [inline src/peft/utils/save_and_load.py:561]: Hmm, I'm unsure about this change. It works here because the `adapter_name` is already part of the key, e.g.: `'base_model.model.lora_head.modules_to_save.default.weight'`. For auxiliary modules, it is being added here:

https://github.com/huggingface/peft/blob/8587a36e3ef490c9313e3c9b622188f511492efc/src/peft/utils/save_and_load.py#L733-L745

Since those modules are already covered, IMO it would make more sense to skip them here completely, WDYT?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for these fixes to how the PEFT state dict is being handled. The error is very niche, I don't think it has a lot of practical examples (if any) but it's still worth fixing.

When I checked, I noticed that the bug is not really tied to only `modules_to_save`, there is a general issue with accidental prefix matches. I added comments regarding that. My general impression is that these fixes are very 'ad hoc' and I'd like to see a more principled solution.

**Comment by BenjaminBossan** [comment]
Heads up, I've started working on a big refactor of the whole `save_and_load.py` module, as it's a bit of a mess right now. This will probably result in fixing the error here as a side effect. It's best to pause the work on this PR until I've gained some clarity.

**Comment by Solaris-star** [comment]
Thanks Benjamin for the detailed review and the heads-up about the refactor.

I'll close this PR to avoid duplicating the fix you're working on in the `save_and_load.py` refactor. The issue #3433 is in good hands — looking forward to seeing the principled solution land.

Happy to review the refactor PR when it's up if an extra pair of eyes helps.

**Comment by BenjaminBossan** [comment]
#3490 was merged. Ideally, you could double check if the latest main branch fixes the issue or if this PR is still required, in which case feel free to re-open it. Either way, thanks for the contribution.
